airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py

```diff
@@ -0,0 +1,337 @@
+            )
+        return None
+
+    @staticmethod
+    def _stream_supports_concurrent_partition_processing(declarative_stream: DeclarativeStream, cursor: DatetimeBasedCursor) -> bool:
+        """
+        Many connectors make use of stream_state during interpolation on a per-partition basis under the assumption that
+        state is updated sequentially. Because the concurrent CDK engine processes different partitions in parallel,
+        stream_state is no longer a thread-safe interpolation context. It would be a race condition because a cursor's
+        stream_state can be updated in any order depending on which stream partition's finish first.
+
+        We should start to move away from depending on the value of stream_state for low-code components that operate
+        per-partition, but we need to gate this otherwise some connectors will be blocked from publishing. See the
+        cdk-migrations.md for the full list of connectors.
+        """
+
+        # If a stream doesn't partition the sync time window into smaller intervals, then we can sync this concurrently because
+        # there is not a race condition on a single partition.
+        if not cursor.step:
+            return True
+
+        if isinstance(declarative_stream.retriever, SimpleRetriever) and isinstance(declarative_stream.retriever.requester, HttpRequester):
+            http_requester = declarative_stream.retriever.requester
+            if "stream_state" in http_requester._path.string:
+                return False
+
+            request_options_provider = http_requester._request_options_provider
+            if request_options_provider.request_options_contain_stream_state():
+                return False
+
+            record_selector = declarative_stream.retriever.record_selector
+            if isinstance(record_selector, RecordSelector):
+                if record_selector.record_filter and "stream_state" in record_selector.record_filter.condition:
+                    return False
+
+                for add_fields in [
+                    transformation for transformation in record_selector.transformations if isinstance(transformation, AddFields)
+                ]:
+                    for field in add_fields.fields:
+                        if isinstance(field.value, str) and "stream_state" in field.value:
+                            return False
+                        if isinstance(field.value, InterpolatedString) and "stream_state" in field.value.string:
+                            return False
+        return True
+
+    @staticmethod
+    def _get_cursor(stream: DeclarativeStream) -> Optional[DeclarativeCursor]:
+        """
+        Returns the low-code cursor component of a stream if it is concurrent compatible. Otherwise, returns None.
+        """
+        if not stream.supports_incremental:
+            return None
+
+        if isinstance(stream.retriever, SimpleRetriever):
```

Comment:
Why not `stream.get_cursor()` here? I fear that this will not work for `AsyncRetriever`

Reply:
yeah this is just an oversight on my part, i've just got rid of this whole method and just call self.get_cursor() from above instead